cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,214 @@
+#include "cpp/merkletree/sparse_merkle_tree.h"
+
+#include <algorithm>
+#include <stddef.h>
+#include <vector>
+
+#include "merkletree/merkle_tree_math.h"
+#include "util/util.h"
+
+using std::make_pair;
+using std::ostream;
+using std::ostringstream;
+using std::reverse;
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+
+vector<string> CalculateNullHashes(const TreeHasher& hasher) {
+  vector<string> ret{hasher.HashLeaf("")};
+  for (int i(1); i < hasher.DigestSize() * 8; ++i) {
+    ret.emplace_back(hasher.HashChildren(ret.back(), ret.back()));
+  }
+  reverse(ret.begin(), ret.end());
+  return ret;
+}
+
+
+Path PathFromBytes(const std::string& bytes) {
+  Path path;
+  for (int i(0); i < path.size(); ++i) {
+    path[i] = (bytes[i / 8] >> (i % 8)) & 0x01;
```

Comment:
How about:
>> (-(i + 1) % 8)
instead of:
>> (i % 8)

That'll give 7, 6, 5, 4, 3, 2, 1, 0, 7, ...

It also seems as though the result of this is being used as the path from the node itself back up through the tree, whereas I'd have expected it to be the other way.  Then again, maybe this makes the inclusion proof code more intuitive?